### PR TITLE
Node lifecycle management

### PR DESCRIPTION
**Pull request by asutula**
This makes some big changes to the way we start/stop the node and query photos. Main changes:

1. The only trigger now is app state transition
2. Location updates do result in app state transitions
3. Created a lock so that only one node management/photo query can be in flight at the same time
4. If the app is launched into the background (because of a location update, for example), the node is started, photos are queried, and then the node is shut down.

Known issue:
Triggers like location change that don't result in an app state transition are ignored right now. This isn't ideal, but also is not very common. For example, if the app is in the foreground and gets a location update, nothing will happen. Similarly, if the app is in a background state and gets a location update, nothing will happen. This case is not ideal, but also the app will often be terminated and launching from terminated into background will trigger the node to start and photos to be queried.

### CODE REVIEW

**Comment by asutula** [comment]
Honestly this stuff is making my head spin. It's complicated. I feel like this PR creates a more stable system, but it's also not the end solution. Need to do some research and thinking and will revisit all this and hopefully create something better.

**Comment by asutula** [comment]
Might have to buy a white board to figure it out hahah

**Comment by andrewxhill** [comment]
going through it now, i'll see if i can understand...

**Comment by andrewxhill** [comment]
checks out, going for it

File: App/Sagas/TextileSagas.js [hunk 1]
```diff
@@ -76,26 +76,42 @@ export function * viewPhoto () {
...
+export function * handleNewAppState ({previousState, newState}) {
+  if (previousState.match(/unknown|background/) && newState === 'background') {
+    console.tron.logImportant('launched into background')
+    yield * triggerCreateNode()
+  } else if (previousState.match(/unknown|inactive|background/) && newState === 'active') {
+    console.tron.logImportant('app transitioned to foreground')
+    yield * triggerCreateNode()
+  } else if (previousState.match(/inactive|active/) && newState === 'background') {
+    console.tron.logImportant('app transitioned to background')
+    yield * triggerStopNode()
```
[the remark this answers is not in the document]
Comment (andrewxhill) [reply App/Sagas/TextileSagas.js]: @carsonfarmer could still be a good place to dive in

[the remark this answers is not in the document]
Comment (carsonfarmer) [reply App/Sagas/TextileSagas.js]: Cool. Digging in now.

[the remark this answers is not in the document]
Comment (carsonfarmer) [reply App/Sagas/TextileSagas.js]: Report back from extensive testing using sims and reactotron:
On iOS, the `launched into background` never seems to fire. Upon toggling the power button (sleep) and waking and unlocking the phone, both the `app transitioned to background` and `app transitioned to foreground` events were fired (and became available in reactotron). `app transitioned to background` *never* seems to fire when (for example) the home button is pressed to background the app.
On android, the `launched into background` event also never seems to fire. However, upon toggling the power button (sleep), `app transitioned to background` fires right away (before sleeping), and `app transitioned to foreground` fires when bringing into foreground from an icon click or from waking up from sleep. Foregrounding and background the app both trigger `app transitioned to background/foreground` right away as expected.
Note: The only real difference is that on iOS, the `app transitioned to background` doesn't fire upon backgrounding the app via the home button, whereas on Android it does right away.

[the remark this answers is not in the document]
Comment (sanderpick) [reply App/Sagas/TextileSagas.js]: good stuff carson

[the remark this answers is not in the document]
Comment (carsonfarmer) [reply App/Sagas/TextileSagas.js]: Ha, Thanks. Next step, try to dig a bit deeper.

**Pull request merged**